e2e/suite_test.go

```diff
@@ -25,93 +24,83 @@ func Test_All(t *testing.T) {
-	}
+	for _, scenario := range scenarioTable {
+		scenario := scenario
+
+		chosenCluster := chooseCompatibleCluster(scenario, clusters)
+		chosenClusterExists := chosenCluster != nil
+		if !chosenClusterExists {
+			t.Logf("could not find test cluster able to run scenario %q, creating a new one...", scenario.Name)
+			newCluster := getBaseClusterModel(
+				fmt.Sprintf(testClusterNameTemplate, randomLowercaseString(r, 5)),
+				suiteConfig.location,
+			)
+			chosenCluster = &newCluster
+			scenario.ScenarioConfig.ClusterMutator(chosenCluster)
+		}
 
-	if err := ensureDebugDaemonset(ctx, kube, suiteConfig.resourceGroupName, suiteConfig.clusterName); err != nil {
-		t.Fatal(err)
-	}
+		if err := ensureCluster(ctx, t, cloud, suiteConfig.location, suiteConfig.resourceGroupName, chosenCluster, !chosenClusterExists); err != nil {
+			t.Fatal(err)
+		}
+
+		clusterName := *chosenCluster.Name
+		t.Logf("chosen cluster name: %q", clusterName)
 
-	var clusterParams map[string]string
-	err = wait.PollImmediateWithContext(ctx, 15*time.Second, 5*time.Minute, func(ctx context.Context) (bool, error) {
-		params, err := extractClusterParameters(ctx, t, kube)
+		subnetID, err := getClusterSubnetID(ctx, cloud, suiteConfig.location, *chosenCluster.Properties.NodeResourceGroup, clusterName)
 		if err != nil {
-			t.Logf("error extracting cluster parameters: %q", err)
-			return false, nil
+			t.Fatal(err)
 		}
-		clusterParams = params
-		return true, nil
-	})
-	if err != nil {
-		t.Fatal(err)
-	}
 
-	if err := createClusterParamsDir(); err != nil {
-		t.Fatal(err)
-	}
+		kube, err := getClusterKubeClient(ctx, cloud, suiteConfig.resourceGroupName, clusterName)
+		if err != nil {
+			t.Fatal(err)
+		}
 
-	t.Logf("dumping cluster parameters to local directory: %s", clusterParamsDir)
-	if err := dumpFileMapToDir(clusterParamsDir, clusterParams); err != nil {
-		t.Error("error dumping cluster parameters", err)
-	}
+		if err := ensureDebugDaemonset(ctx, kube); err != nil {
+			t.Fatal(err)
+		}
 
-	baseConfig, err := getBaseBootstrappingConfig(ctx, t, cloud, suiteConfig, clusterParams)
-	if err != nil {
-		t.Fatal(err)
-	}
+		clusterParams, err := pollExtractClusterParameters(ctx, t, kube)
+		if err != nil {
+			t.Fatal(err)
+		}
 
-	for name, tc := range cases {
-		if suiteConfig.testsToRun != nil && !suiteConfig.testsToRun[name] {
-			continue
+		baseConfig, err := getBaseNodeBootstrappingConfiguration(ctx, t, cloud, suiteConfig, clusterParams)
+		if err != nil {
+			t.Fatal(err)
 		}
 
-		tc := tc
-		caseName := name
 		copied, err := deepcopy.Anything(baseConfig)
 		if err != nil {
 			t.Error(err)
 			continue
 		}
 		nbc := copied.(*datamodel.NodeBootstrappingConfiguration)
 
-		if tc.bootstrapConfigMutator != nil {
-			tc.bootstrapConfigMutator(t, nbc)
+		if scenario.ScenarioConfig.BootstrapConfigMutator != nil {
+			scenario.ScenarioConfig.BootstrapConfigMutator(t, nbc)
 		}
 
-		t.Run(caseName, func(t *testing.T) {
+		t.Run(scenario.Name, func(t *testing.T) {
```

Comment:
our run function is also getting a little long -- maybe extract this all to a `runScenario` or similar helper?

Reply:
will do